demo/lib/distro.bash

```diff
@@ -0,0 +1,339 @@
+GOLANG_VERSION=1.14.8
+GOLANG_URLDIR=https://dl.google.com/go
+
+# distro-specific function resolution
+distro-resolve() {
+    local fn fallback
+    case $VM_DISTRO in
+        ubuntu*) fallback="debian-$1";;
+        centos*) fallback="fedora-$1 rpm-$1";;
+        fedora*) fallback="rpm-$1";;
+        *suse*)  fallback="rpm-$1";;
+    esac
+    # try version-based resolution first, then derivative fallback
+    for fn in "${VM_DISTRO/./_}-$1" "${VM_DISTRO%-*}-$1" $fallback; do
+        if [ "$(type -t -- $fn)" = "function" ]; then
+            echo $fn
+            return 0
+        fi
+    done
+    echo "UNRESOLVABLE-$1-FOR-$VM_DISTRO"
+    return 1
+}
+
+###########################################################################
+
+#
+# distro-agnostic interface
+#
+# To add a new distro implement distro-specific versions of these
+# functions. You can omit implementing those which already resolve
+# to an existing function which works for the new distro.
+#
+
+distro-image-url() {
+    local fn=$(distro-resolve ${FUNCNAME#distro-})
+    $fn
+}
+
+distro-ssh-user() {
+    local fn=$(distro-resolve ${FUNCNAME#distro-})
+    $fn
+}
```

Comment:
> To avoid repeating this code, how about defining all functions in a loop:
> 
> ```
> for func in \
>     distro-image-url \
>     distro-ssh-user \
>    ...
>     ; do eval "
> $func() {
>     local fn=\$(distro-resolve \${FUNCNAME#distro-})
>     \$fn \"\$@\"
> }
> "; done
> ```

I'm not sure about that. I think it would just make it more difficult to read, but that is admittedly a subjective preference.
Instead I've been considering changing `distro-resolve` so that it would
  - use `${FUNCNAME[1]}` (IOW the calling function's name) implicitly as the function to resolv, and
  - resolve and then invoke the resulting function passing on `"$@"` (now the templates are not identical in this sense)

With such changes the distro interface section would become a set of simple trivially copy-pastable one-liners:

```
distro-resolve() {
   local caller="${FUNCNAME[1]}" name="${caller#distro-}" fn
   ...
   # resolve name to fn
   ...
   $fn "$@"
}

distro-image-url()          { distro-resolve "$@"; }
distro-ssh-user()           { distro-resolve "$@"; }
distro-pkg-type()           { distro-resolve "$@"; }
...
distro-set-kernel-cmdline() { distro-resolve "$@"; }
```

WDYT?

Reply:
Yeah, that's more friendly for eyes.

I'd probably still pass the function name as the first argument to the resolver, so it wouldn't depend on the name of the calling function.

```
distro-image-url()          { distro-resolve "${FUNCNAME}" "$@"; }
distro-ssh-user()           { distro-resolve "${FUNCNAME}" "$@"; }
...
```